src/scripts/redirects.mjs

```diff
@@ -1,18 +1,43 @@
 import { getMovedPagesFromHistory } from './getMovedPagesFromHistory.mjs'
 
+/**
+ * Cleans up and filters a list of moved files — removing cycles,
+ * resolving redirect chains, etc.
+ *
+ * @param {{ source: string, destination: string }[]} movedPages
+ * @returns {{ source: string, destination: string }[]}
+ */
+function cleanupRedirects(movedPages) {
+    // Build a map of the redirects, so that we can detect cycles and resolve
+    // redirect chains.
+    const redirects = new Map()
+    for (const { source, destination } of movedPages) {
+        const nextRedirect = redirects.get(destination)
+        if (nextRedirect !== source) {
+            const finalDestination = nextRedirect || destination
+            redirects.set(source, finalDestination)
+        }
+        // If nextRedirect === source, it's a direct loop — drop it entirely.
+    }
+    return [...redirects.entries()].map(([source, destination]) => ({ source, destination }))
+}
+
+export const cleanupRedirectsForTesting = cleanupRedirects
+
 /**
  * Returns all redirects that should be generated.
  *
  * @returns {Promise<{ source: string, destination: string }[]>}
  */
 export default async function redirects() {
+    const movedPages = await getMovedPagesFromHistory()
     return [
-        ...await getMovedPagesFromHistory(),
+        ...cleanupRedirects(movedPages),
```

Comment:
Should we call cleanupRedirects on the entire array, instead of just movedPages? That way we can catch instances of a custom redirect causing a redirect loop as well

Reply:
Done.